Implement public API InferenceMode and its error handling

ailzhang opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/53343

For easier review, here's a diff between the version before revert. https://www.internalfb.com/phabricator/paste/view/P363902707

Test Plan: Imported from OSS

Differential Revision: D27443229

Pulled By: ailzhang

ailzhang commented:
Changes compared to the version before reland: https://gist.github.com/ailzhang/992728d6964afd86a42ea56c6e43835b

bhosmer commented on c10/core/DispatchKeySet.h:
```diff
@@ -203,6 +207,11 @@ constexpr DispatchKeySet autograd_dispatch_keyset = DispatchKeySet({
 });
 
+// See Note [TLS Initialization]
+constexpr DispatchKeySet InplaceOrView_keyset = DispatchKeySet({
```
What if we called this `default_included_keyset` and described our scheme in more general terms? Like, "To preserve the invariant that raw TLS storage of the default state is zero, we obtain the *actual* current include keyset by xoring raw storage with the default state defined here". It's actually not a bad setup, and framing it this way would clarify how to add more nonzero default bits in the future if necessary (as well as using the same recipe if we needed other nonzero initial TLS values).

bhosmer commented on c10/core/InferenceMode.cpp:
```diff
@@ -0,0 +1,10 @@
+namespace c10 {
+
+bool InferenceMode::is_enabled() {
+  return c10::impl::tls_is_dispatch_key_included(DispatchKey::InplaceOrView);
```
This is gonna be hella confusing without a comment, I think. I'd even think about reverting this back to `!` and making `is_dispatch_keyset_included` do the xor, i.e. encapsulate the new technique. But if you want to avoid the extra instruction or whatever, I'd consider at least renaming to `is_dispatch_keyset_included_in_raw_storage` to clarify what the function is actually computing...

ailzhang replied:
I actually have a note in InferenceMode.h explaining the expected state of InplaceOrView in InferenceMode which I think can explain it. I'll add a link to note here! :D

bhosmer commented on c10/core/impl/LocalDispatchKeySet.cpp:
```diff
@@ -5,7 +5,12 @@
+// We want raw_local_dispatch_key_set to be initialized with InplaceOrView key
+// in the included set. But certain Windows compiler (e.g the one used in ARVR tests)
+// only allow TLS to be zero-initialized. So we're working around the problem by adding
+// XOR raw_local_dispatch_key_set.included() with InplaceOrView in DispatchKeyExtractor.h.
```
Here's where I think we can describe the scheme in more general terms, rather than strictly a workaround for `InplaceOrView`.

bhosmer reviewed (commented):
THANKS for the diff from the earlier version! 😁  This LGTM but interested in your take on a) framing the scheme a little more generally and b) encapsulating the xor so callsites don't need to remember to do it etc. (per inline comments).

Commit:
Implement public API InferenceMode and its error handling (#55008)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/55008

reland of https://github.com/pytorch/pytorch/pull/53343

For easier review, here's a diff between the version before revert. https://www.internalfb.com/phabricator/paste/view/P361764610

Test Plan: Imported from OSS

Differential Revision: D27443229

Pulled By: ailzhang

fbshipit-source-id: faeaff3b6165b933c9f354d5f0344e38269fbb12

bhosmer reviewed (approved):
Thanks!
